db/fetcher/indexer_iterators.go

```diff
@@ -274,108 +274,75 @@ func (iter *scanningIndexIterator) Next() (indexIterResult, error) {
 
 // checks if the value satisfies the condition
 type valueMatcher interface {
-	Match(any) (bool, error)
+	Match(client.NormalValue) (bool, error)
 }
 
 type intMatcher struct {
 	value    int64
 	evalFunc func(int64, int64) bool
 }
 
-func (m *intMatcher) Match(value any) (bool, error) {
-	intVal, ok := value.(int64)
-	if !ok {
-		return false, NewErrUnexpectedTypeValue[int64](value)
+func (m *intMatcher) Match(value client.NormalValue) (bool, error) {
+	intVal, ok := value.Int()
+	if ok {
+		return m.evalFunc(intVal, m.value), nil
+	}
+	intOptVal, ok := value.NillableInt()
+	if ok {
+		return m.evalFunc(intOptVal.Value(), m.value), nil
```

Comment:
question: This looks like a potential behaviour change to me, we are not checking for `nil` here - is that alright here?

suggestion: Maybe a func like the below would be handy (annoyingly generic funcs cant be defined on Go interfaces or we could replace most funcs on it with a couple of generic ones...):
```
func AsT[T client.NormalValueConstraints](v NormalValue) (T, bool) {
  any := v.Unwrap()
  return any.(T)
}
```
It could either live here as a private func, or in the client package, then in the funcs here you wouldn't need two ifs, you could just:
```
func (m *intMatcher) Match(value client.NormalValue) (bool, error) 
  intVal, ok := value.AsT[int64]()
  if ok {
    return m.evalFunc(intVal, m.value), nil
  }
  return false, NewErrUnexpectedTypeValue[int64](value)
}
```

Reply:
> annoyingly generic funcs cant be defined on Go interfaces or we could replace most funcs on it with a couple of generic ones...

They can be defined on interfaces: https://go.dev/play/p/gt5p6hQsq_Q

Reply:
> They can be defined on interfaces:

Only if the type param is declared on the interface, which you dont want to do here.

Reply:
I don't a benefit here. Why can't you just write `v, ok := value.Unwrap().(int64)`?

Reply:
> Why can't you just write v, ok := value.Unwrap().(int64)?

That is my suggestion, only wrapped up into a func so you don't have to cast in a whole bunch of places.

Reply:
`v, ok := value.Unwrap().(int64)` this wouldn't work as it wouldn't check for `nil` passing `0` to the `evalFunc` is no a correct behaviour if the actual value is `nil`.

Reply:
Ah true, the private func would need to be:
```
func AsT[T client.NormalValueConstraints](v NormalValue) (T, bool) {
  if v.IsNil() {
    return false
  }
  return v.Unwrap().(T), true
}
```

Reply:
I just changed code to check it explicitly. You suggestion wouldn't work as it will return 0 for `nil` values which will match if the user runs a filter `_eq: 0`